lib/pipelines/code-pipeline.ts

```diff
@@ -2,57 +2,166 @@ import * as cdk from '@aws-cdk/core';
 import * as actions from '@aws-cdk/aws-codepipeline-actions';
+import { Construct, StackProps } from '@aws-cdk/core';
+import { ApplicationRepository, AsyncStackBuilder, StackBuilder } from '../spi';
+import { AddStageOptions } from '@aws-cdk/pipelines';
+import { withUsageTracking } from '../utils/usage-utils'; 
+
+
+/**
+ * credentialsType is excluded and the only supported credentialsSecret is a plaintext GitHub OAuth token.
+ */
+export type GitHubSourceRepository = Omit<ApplicationRepository, "credentialsType">;
 
 /**
  * Props for the Pipeline.
  */
 export type PipelineProps = {
+    
     /**
      * The name for the pipeline.
      */
-    name: string
+    name: string;
 
     /**
      * The owner of the repository for the pipeline (GitHub handle).
      */
-    owner: string
+    owner: string;
+
+    /**
+     * Repository for the pipeline
+     */
+    repository: GitHubSourceRepository;
 
     /**
-     * The name of the repository for the pipeline.
+     * Pipeline stages and options.
      */
-    repo: string
+    stages: StackStage[];
+}
 
+/**
+ * Stack stage is a builder construct to allow adding stages to a pipeline. Each stage is expected to produce a stack. 
+ */
+export interface StackStage {
     /**
-     * The branch for the pipeline.
+     * id of the stage
      */
-    branch: string
+    id: string;
 
     /**
-     * Secret key for GitHub OAuth credentials (stored in SecretsManager).
+     * Builder that can produce a stack which will be deployed as part of the stage
      */
-    secretKey?: string
+    stackBuilder: StackBuilder;
 
     /**
-     * The CDK scope for the pipeline.
+     * Optional stage properties, such as {manualApprovals: true} which can control stage transitions.
      */
-    scope: cdk.Construct
+    stageProps?: AddStageOptions
 }
 
 /**
- * CodePipeline deploys a new CodePipeline resource that is integrated with a GitHub repository.
+ * Builder for CodePipeline.
  */
-export class CodePipeline {
+export class CodePipelineBuilder implements StackBuilder {
+
+    private props: Partial<PipelineProps>;
 
-    public static build = (props: PipelineProps) => {
+
+    constructor() {
+        this.props = { stages: []};
+    }
+
+    public name(name: string): CodePipelineBuilder {
+        this.props.name = name;
+        return this;
+    }
+
+    public owner(owner: string) : CodePipelineBuilder {
+        this.props.owner = owner;
+        return this;
+    }
+
+    public repository(repo: GitHubSourceRepository): CodePipelineBuilder {
+        this.props.repository = repo;
+        return this;
+    }
+
+    public stage(stackStage: StackStage) : CodePipelineBuilder {
+        this.props.stages?.push(stackStage);
+        return this;
+    }
+    
+    build(scope: cdk.Construct, id: string, stackProps?: cdk.StackProps): cdk.Stack {
+        console.assert(this.props.name && this.props.owner && this.props.repository!.credentialsSecretName && this.props.stages, 
```

Comment:
consider asserting on each individually. Helps be specific about which property is missing (and avoids a super long line of code!)

Reply:
Addressed